compiler/rustc_ty_utils/src/needs_drop.rs

```diff
@@ -195,25 +191,25 @@ where
                     // for the coroutine witness and check whether any of the contained types
                     // need to be dropped, and only require the captured types to be live
                     // if they do.
-                    ty::Coroutine(_, args) => {
-                        if self.reveal_coroutine_witnesses {
-                            queue_type(self, args.as_coroutine().witness());
+                    ty::Coroutine(def_id, args) => {
+                        if self.exhaustive {
+                            for upvar in args.as_coroutine().upvar_tys() {
+                                queue_type(self, upvar);
+                            }
+                            queue_type(self, args.as_coroutine().resume_ty());
+                            if let Some(witness) = tcx.mir_coroutine_witnesses(def_id) {
+                                for field_ty in &witness.field_tys {
+                                    queue_type(
+                                        self,
+                                        EarlyBinder::bind(field_ty.ty).instantiate(tcx, args),
+                                    );
+                                }
+                            }
                         } else {
+                            // TODO: We could still recurse into upvars and the resume here.
```

Comment:
This is definitely a FIXME for later. I don't think this has any side effect.

Reply:
I don't think we should :grin: you could even inline `always_drop_component` to just return `Err(AlwaysRequiresDrop)` here :grin: